src/poprox_recommender/pipeline.py

```diff
@@ -71,14 +77,78 @@ def run_component(self, component_spec: ComponentSpec, state: PipelineState):
-        if not isinstance(output, (ArticleSet, InterestProfile)):
+        self._validate_return_type(component_spec, output)
+        state._last = component_spec.output
+        state[state._last] = output
+
+        return state
+
+    def _validate_input_types(self, spec: ComponentSpec, state_types: dict[str, type]):
+        # Best effort type-checking to highlight when inputs are mismatched or out of order
+        sig = signature(spec.component.__call__)
+        sig_params = list(sig.parameters.values())
+
+        for input_name, sig_param in zip(spec.inputs, sig_params):
+            input_type = state_types[input_name]
+            sig_param_type = sig_param.annotation
+            if not included_in(input_type, sig_param_type):
+                msg = (
+                    f"Component {spec.component} expected inputs with types {[p.annotation for p in sig_params]} "
+                    f"but received inputs with types {[state_types[i] for i in spec.inputs]}"
+                )
+
+                raise TypeError(msg)
+
+    def _validate_output_type(self, spec: ComponentSpec):
+        # Best effort type-checking to highlight when output is
+        # mismatched with existing state values types
+        sig = signature(spec.component.__call__)
+        output_name = spec.output
+
+        state_type = self._state_types.get(output_name, None)
+        output_type = sig.return_annotation
+
+        if output_type is not _empty and output_type not in (ArticleSet, InterestProfile):
             msg = (
                 f"Pipeline components must return ArticleSet or InterestProfile, "
-                f"but received {type(output)} from {type(component_spec.component)}"
+                f"but received {type(output_type)} from {type(spec.component)}"
             )
             raise TypeError(msg)
 
-        state._last = component_spec.output
-        state[state._last] = output
+        if output_type is not _empty and not included_in(output_type, state_type):
+            msg = (
+                f"Component {spec.component} returns output with type {output_type} "
+                f"but would need to return {state_type} in order to overwrite {output_name}"
+            )
 
-        return state
+            raise TypeError(msg)
+
+        return output_type
+
+    def _validate_return_type(self, component_spec, output):
+        expected_type = self._state_types.get(component_spec.output, None)
+        output_type = type(output)
+        if not included_in(output_type, expected_type):
+            msg = (
+                f"{type(component_spec.component)} is expected to return {expected_type}, "
+                f"but received {type(output)}"
+            )
+            raise TypeError(msg)
+
+
+def is_union(t: Any) -> bool:
+    origin = get_origin(t)
+    return origin is Union or origin is UnionType
+
+
+def included_in(object_type: type, t: Any):
+    if t == _empty or t is None:
+        return True
+
+    if object_type == t:
```

Comment:
Should this also have `issubclass`, possibly gated on a `covariant` boolean argument? If A subclasses B, then an output of A is valid for an input of B.

Reply:
Do we expect to subclass `ArticleSet`, `InterestProfile`, or other types we'd use this on? No objection to improving the robustness of this, just wondering how much this is a theoretical concern vs a practical problem.

Reply:
I think subclassing is unlikely in practice, more of a theoretical correctness concern (and a "get the type logic right now, b/c types are a place where theoretical correctness is important, and so we don't have to page this code back into working memory later if that situation changes").